---
title: "Exploring the single-cell RNA-seq analysis landscape"
author:
- Luke Zappia
- Belinda Phipson
- Alicia Oshlack
date: "`r Sys.Date()`"
output:
    bookdown::word_document2:
        reference_docx: style/style.docx
bibliography: style/references.bib
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(autodep        = TRUE,
                      cache          = FALSE,
                      cache.path     = "cache/",
                      cache.comments = TRUE,
                      echo           = FALSE,
                      error          = FALSE,
                      fig.path       = "figures/",
                      fig.width      = 10,
                      fig.height     = 8,
                      dev            = c('png', 'pdf'),
                      message        = FALSE,
                      warning        = FALSE)
```

```{r libaries}
# Plotting
library("cowplot")
library("RColorBrewer")

# Presentation
library("knitr")

# Strings
library("stringr")

# JSON
library("jsonlite")

# Tidyverse
library("tidyverse")
```

```{r palette}
pal <- c(pink = "#EC008C",    # Pink
         blue = "#00ADEF",    # Blue
         green = "#8DC63F",   # Green
         teal = "#00B7C6",    # Teal
         orange = "#F47920",  # Orange
         purple = "#7A52C7")  # Purple
```

```{r load}
tools <- read_csv("data/single_cell_software.csv",
                  col_types = cols(
                      .default = col_logical(),
                      Name = col_character(),
                      Platform = col_character(),
                      DOI = col_character(),
                      PubDate = col_character(),
                      Code = col_character(),
                      Description = col_character(),
                      License = col_character(),
                      Added = col_date(format = ""),
                      Updated = col_date(format = "")
                  )) %>%
    mutate(IsOld = Added < "2016-10-01")

cats <- read_json("data/descriptions.json", simplifyVector = TRUE) %>%
    mutate(Phase = factor(Phase,
                          levels = c("Phase 1", "Phase 2", "Phase 3", "Phase 4",
                                     "Multiple", "Other")))
```

# Abstract

# Keywords

# Introduction

* scRNA-seq
    * What is it?
    * What is the history?
    * What is it used for?
    * How is it different to bulk?
* Analysis
    * Getting the data
    * Quality control
    * Clustering/psuedotime
    * Gene identification
    * Interpretation?

Single-cell RNA-sequencing (scRNA-seq) has rapidly gained traction as an
effective tool for interrogating the transcriptome at the resolution of
individual cells. Since the first protocols were published in 2009 the number of
cells profiled in individual scRNA-seq experiments has increased exponentially,
outstripping Mooreâ€™s Law [Svensson 2017]. This new kind of transcriptomic data
brings a demand for new analysis methods. Not only does the scale of scRNA-seq
datasets vastly outstrip bulk experiments but there are also a variety of
challenges specific to the single-cell context. Specifically, scRNA-seq data is
extremely sparse (there is no expression measured for many genes in most cells),
contains large technical artifacts and the scientific questions of interest are
often different to those asked of bulk RNA-seq datasets. For example many bulk
RNA-seq datasets are generated to detect differentially expressed genes through
a designed experiment while many scRNA-seq experiments aim to identify or
classify cell types.

The bioinformatics community has also embraced this new type of data with a
plethora of methods for the analysis of scRNA-seq data being produced. As such,
keeping up with the current state of scRNA-seq analysis is now a significant
challenge as field is presented with a huge number of choices for approaching an
analysis. Since September 2016 we have collated and categorised scRNA-seq
analysis tools as they have become available. This database is being continually
updated and is publicly available at www.scRNA-tools.org. In order to help
researchers navigate the analysis jungle we discuss the stages of scRNA-seq
analysis and their relationship to tools and categories in the scRNA-tools
database. Through the analysis of this database we show trends in not only the
rate of publication of tools but also the analysis application their methods
address. Based on this database we gain insight into the trends in this rapidly
developing field.

# Overview of the scRNA-tools database

* What is in the database?
* How has it been collected?

The scRNA-tools database contains information on software tools specifically
designed for the analysis of scRNA-seq data. For a tool to be eligible to be
included in the database it must be available for download and public use. This
can be from a software package repository (such as Bioconductor, CRAN or PyPI),
a code sharing website such as Github or directly from a private website.
Various details of the tools are recorded such as the programming language or
platform they use, details of any related publication, links to the source code
and the associated software license. Tools are also categorised according to the
analysis tasks they are able to perform. Most tools are added after a preprint
or publication becomes available but some have been added after being mentioned
on social media or in similar collections.

When the database was first constructed there were 70 scRNA-seq analysis tools
available, representing work in the field during the three years from the first
published tool in November 2013 [SAMstrt] up to October 2016. In the year since
then over 60 more tools have been added [FIGURE]. The doubling of the number of
available tools in such a short time demonstrates the blossoming interest in
scRNA-seq and its maturation from a technique requiring custom-built equipment
with specialised protocols to a commercially available product.

* Increase in tools over time

```{r plot-tools-over-time}
date.totals <- tools %>%
    select(Date = Added) %>%
    group_by(Date = as.Date(Date)) %>%
    summarise(Count = n()) %>%
    mutate(Total = cumsum(Count))

totals.plot <- ggplot(date.totals, aes(x = Date, y = Total)) +
    geom_line(size = 2, colour = pal["purple"]) +
    xlab("Date") +
    ylab("Number of tools") +
    ggtitle("Increase in tools over time") +
    theme_cowplot() +
    theme(plot.title = element_text(size = 20),
          axis.text = element_text(size = 12)
    )

ggsave("figures/tools_over_time.png", totals.plot, height = 5)
ggsave("figures/tools_over_time.pdf", totals.plot, height = 5)
```

```{r tools-over-time, fig.cap = "Number of tools in the scRNA-tools database over time. Since the scRNA-seq tools database was started in September 2016 more than 60 new tools have been released."}
include_graphics("figures/tools_over_time.png")
```

## Publication status

* Publication status
* ~ half published, more as preprints
* More older tools published, reflects publishing delay

```{r plot-publication-status}
pub.plot <- tools %>%
    select(PubDate) %>%
    mutate(IsPre = PubDate == "PREPRINT") %>%
    mutate(IsPub = !is.na(PubDate) & !IsPre) %>%
    mutate(IsNot = is.na(PubDate)) %>%
    summarise(NotPublished = sum(IsNot),
              Published = sum(IsPub, na.rm = TRUE),
              Preprint = sum(IsPre, na.rm = TRUE)) %>%
    gather(key = Type, value = Count) %>%
    mutate(Type = factor(Type,
                         levels = c("Preprint", "Published", "NotPublished"),
                         labels = c("Preprint", "Published", "Not Published")),
           Cumulative = cumsum(Count),
           Midpoint = Cumulative - (Count / 2),
           Label = paste0(Type, "\n",
                          round(Count / sum(Count) * 100, 1), "%")) %>%
    ggplot(aes(x = 1, weight = Count, fill = Type)) +
    geom_bar(width = 1, position = "stack") +
    coord_polar(theta = "y") +
    geom_text(aes(x = 1.8, y = Midpoint, label = Label, colour = Type),
              size = 6) +
    scale_fill_manual(values = unname(pal)) +
    scale_colour_manual(values = unname(pal)) +
    ggtitle("Publication status") +
    theme_nothing() +
    theme(plot.title = element_text(size = 20),
          legend.position = "none"
    )

pub.plot.time <- tools %>%
    group_by(IsOld) %>%
    select(PubDate, IsOld) %>%
    mutate(IsPre = PubDate == "PREPRINT") %>%
    mutate(IsPub = !is.na(PubDate) & !IsPre) %>%
    mutate(IsNot = is.na(PubDate)) %>%
    summarise(NotPublished = sum(IsNot),
              Published = sum(IsPub, na.rm = TRUE),
              Preprint = sum(IsPre, na.rm = TRUE)) %>%
    gather(key = Type, value = Count, -IsOld) %>%
    group_by(IsOld) %>%
    mutate(Prop = Count / sum(Count)) %>%
    mutate(Type = factor(Type,
                         levels = c("Preprint", "Published", "NotPublished"),
                         labels = c("Preprint", "Published", "Not Published"))) %>%
    ggplot(aes(x = Type, y = Prop, fill = IsOld)) +
    geom_bar(stat = "identity", position = "dodge") +
    scale_fill_manual(name = "Date added",
                      labels = c("After 2016-10-01", "Before 2016-10-01"),
                      values = unname(pal)) +
    scale_y_continuous(labels = scales::percent) +
    ggtitle("Publication status over time") +
    ylab("Percentage of tools") +
    theme_cowplot() +
    theme(legend.position = c(0.75, 0.85),
          legend.title = element_text(size = 14),
          legend.text = element_text(size = 12),
          legend.key.size = unit(25, "points"),
          plot.title = element_text(size = 20),
          axis.text = element_text(size = 12),
          axis.title.x = element_blank()
    )

panel <- plot_grid(pub.plot, pub.plot.time, nrow = 1, ncol = 2,
                   labels = "AUTO")

save_plot("figures/publication_status.png", panel, nrow = 1, ncol = 2,
          base_height = 8)
save_plot("figures/publication_status.pdf", panel, nrow = 1, ncol = 2,
          base_height = 8)
```

```{r publication-status, fig.cap = "Publication status of tools in the scRNA-tools database. Over half of the tools in the full database have been published in peer-revewied journals while another third are available as preprints (A). When stratified by publication date we see that the majority of tools added before October 2016 are published, while the majority of newer tools are available as preprints (B). The number of unpublished tools has stayed consistent at around 10 percent."}
include_graphics("figures/publication_status.png")
```

```{r plot-licenses}
license.plot <- tools %>%
    select(License) %>%
    mutate(IsGPL = str_detect(License, "GPL"),
           IsBSD = str_detect(License, "BSD"),
           IsMIT = str_detect(License, "MIT"),
           IsApache = str_detect(License, "Apache"),
           IsArtistic = str_detect(License, "Artistic"),
           IsUnknown = is.na(License),
           IsOther = !(IsGPL | IsBSD | IsMIT | IsApache | IsArtistic |
                           IsUnknown)) %>%
    summarise(Apache = sum(IsApache, na.rm = TRUE),
              Artistic = sum(IsArtistic, na.rm = TRUE),
              BSD = sum(IsBSD, na.rm = TRUE),
              GPL = sum(IsGPL, na.rm = TRUE),
              MIT = sum(IsMIT, na.rm = TRUE),
              Other = sum(IsOther),
              Unknown = sum(IsUnknown)) %>%
    gather(key = License, value = Count) %>%
    mutate(License = factor(License,
                            levels = c("Apache", "Artistic", "BSD", "GPL",
                                       "MIT", "Other", "Unknown")),
           Cumulative = cumsum(Count),
           Midpoint = Cumulative - (Count / 2),
           Label = paste0(License, "\n",
                          round(Count / sum(Count) * 100, 1), "%")) %>%
    ggplot(aes(x = 1, weight = Count, fill = License)) +
    geom_bar(width = 1, position = "stack") +
    coord_polar(theta = "y") +
    geom_text(aes(x = 1.75, y = nrow(tools) - Midpoint, label = Label,
                  colour = License), size = 6) +
    scale_fill_manual(values = c(unname(pal), "#999999")) +
    scale_colour_manual(values = c(unname(pal), "#999999")) +
    ggtitle("Software licenses") +
    theme_nothing() +
    theme(plot.title = element_text(size = 20),
          legend.position = "none"
    )

license.plot.time <- tools %>%
    group_by(IsOld) %>%
    select(License, IsOld) %>%
    mutate(IsGPL = str_detect(License, "GPL"),
           IsBSD = str_detect(License, "BSD"),
           IsMIT = str_detect(License, "MIT"),
           IsApache = str_detect(License, "Apache"),
           IsArtistic = str_detect(License, "Artistic"),
           IsUnknown = is.na(License),
           IsOther = !(IsGPL | IsBSD | IsMIT | IsApache | IsArtistic |
                           IsUnknown)) %>%
    summarise(Apache = sum(IsApache, na.rm = TRUE),
              Artistic = sum(IsArtistic, na.rm = TRUE),
              BSD = sum(IsBSD, na.rm = TRUE),
              GPL = sum(IsGPL, na.rm = TRUE),
              MIT = sum(IsMIT, na.rm = TRUE),
              Other = sum(IsOther),
              Unknown = sum(IsUnknown)) %>%
    gather(key = License, value = Count, -IsOld) %>%
    mutate(License = factor(License,
                            levels = c("Apache", "Artistic", "BSD", "GPL",
                                       "MIT", "Other", "Unknown"))) %>%
    group_by(IsOld) %>%
    mutate(Prop = Count / sum(Count)) %>%
    ggplot(aes(x = License, y = Prop, fill = IsOld)) +
    geom_bar(stat = "identity", position = "dodge") +
    scale_fill_manual(name = "Date added",
                      labels = c("After 2016-10-01", "Before 2016-10-01"),
                      values = unname(pal)) +
    scale_y_continuous(labels = scales::percent) +
    ggtitle("Software licenses") +
    ylab("Percentage of tools") +
    theme_cowplot() +
    theme(legend.position = c(0.75, 0.85),
          legend.title = element_text(size = 14),
          legend.text = element_text(size = 12),
          legend.key.size = unit(25, "points"),
          plot.title = element_text(size = 20),
          axis.text = element_text(size = 12),
          axis.title.x = element_blank()
    )

panel <- plot_grid(license.plot, license.plot.time, nrow = 1, ncol = 2,
                   labels = "AUTO")

save_plot("figures/licenses.png", panel, nrow = 1, ncol = 2,
          base_height = 8)
save_plot("figures/licenses.pdf", panel, nrow = 1, ncol = 2,
          base_height = 8)
```

```{r licenses, eval = FALSE}
include_graphics("figures/licenses.png")
```

```{r plot-platforms}
platforms.plot <- tools %>%
    select(Platform) %>%
    mutate(IsR = str_detect(Platform, "R"),
           IsPython = str_detect(Platform, "Python"),
           IsMATLAB = str_detect(Platform, "MATLAB"),
           IsCPP = str_detect(Platform, "C++"),
           IsOther = !(IsR | IsPython | IsMATLAB | IsCPP)) %>%
    summarise(R = sum(IsR),
              Python = sum(IsPython),
              MATLAB = sum(IsMATLAB),
              CPP = sum(IsCPP),
              Other = sum(IsOther)) %>%
    gather(key = Platform, value = Count) %>%
    mutate(Platform = factor(Platform,
                             levels = c("R", "Python", "MATLAB", "CPP",
                                        "Other"),
                             labels = c("R", "Python", "MATLAB", "C++",
                                        "Other"))) %>%
    ggplot(aes(x = Platform, weight = Count, fill = Platform)) +
    geom_bar(aes(y = (..count..) / sum(..count..))) +
    scale_y_continuous(labels = scales::percent) +
    scale_fill_manual(values = unname(pal)) +
    ylab("Percentage of tools") +
    ggtitle("Platforms used by analysis tools") +
    theme_cowplot() +
    theme(axis.title.x = element_blank(),
          legend.position = "none",
          plot.title = element_text(size = 20),
          axis.text = element_text(size = 12)
    )

platforms.plot.time <- tools %>%
    group_by(IsOld) %>%
    select(Platform, IsOld) %>%
    mutate(IsR = str_detect(Platform, "R"),
           IsPython = str_detect(Platform, "Python"),
           IsMATLAB = str_detect(Platform, "MATLAB"),
           IsCPP = str_detect(Platform, "C++"),
           IsOther = !(IsR | IsPython | IsMATLAB | IsCPP)) %>%
    summarise(R = sum(IsR),
              Python = sum(IsPython),
              MATLAB = sum(IsMATLAB),
              CPP = sum(IsCPP),
              Other = sum(IsOther)) %>%
    gather(key = Platform, value = Count, -IsOld) %>%
    mutate(Platform = factor(Platform,
                             levels = c("R", "Python", "MATLAB", "CPP",
                                        "Other"),
                             labels = c("R", "Python", "MATLAB", "C++",
                                        "Other"))) %>%
    group_by(IsOld) %>%
    mutate(Prop = Count / sum(Count)) %>%
    ggplot(aes(x = Platform, weight = Prop, fill = IsOld)) +
    geom_bar(position = "dodge") +
    scale_y_continuous(labels = scales::percent) +
    scale_fill_manual(name = "Date added",
                      labels = c("After 2016-10-01", "Before 2016-10-01"),
                      values = unname(pal)) +
    ggtitle("Platforms used by analysis tools") +
    ylab("Percentage of tools") +
    theme_cowplot() +
    theme(legend.position = c(0.75, 0.85),
          legend.title = element_text(size = 14),
          legend.text = element_text(size = 12),
          legend.key.size = unit(25, "points"),
          plot.title = element_text(size = 20),
          axis.text = element_text(size = 12),
          axis.title.x = element_blank()
    )

panel <- plot_grid(platforms.plot, platforms.plot.time, nrow = 1, ncol = 2,
                   labels = "AUTO")

save_plot("figures/platforms.png", panel, nrow = 1, ncol = 2,
          base_height = 8)
save_plot("figures/platforms.pdf", panel, nrow = 1, ncol = 2,
          base_height = 8)
```

```{r platforms, eval = FALSE}
include_graphics("figures/platforms.png")
```

```{r plot-platforms-licenses}
panel <- plot_grid(platforms.plot, license.plot, nrow = 1, ncol = 2,
                   labels = "AUTO")

save_plot("figures/platforms_licenses.png", panel, nrow = 1, ncol = 2,
          base_height = 8)
save_plot("figures/platforms_licenses.pdf", panel, nrow = 1, ncol = 2,
          base_height = 8)
```

## Open-source

* Most tools use (free) R or Python
* Most released under open-source license
* Some with no license...

Developers of scRNA-seq analysis tools have choices to make about what platforms
they use to create their tools, how they make them available to the community
and whether they share the source code. We find that the most commonly used
platform is the R statistical programming language, with many tools being
available through the Bioconductor or CRAN repositories. Python is the second
most popular language, followed by MATLAB, a commercially available product, and
the lower-level C++. The use of R and Python is consistent with their popularity
across a range of data science fields. In particular the use of R reflects its
history as the language of choice for the analysis of bulk RNA-seq datasets and
a range of other biological data types.

```{r platforms-licenses, fig.cap = "The open-source nature of scRNA-seq analysis tools. The majority of tools are available for either the R or Python programming languages (A). Most tools are released under a standard open-source software license, with variants of the GNU Public License (GPL) being the most common. However licenses could not be found for a large proportion of tools."}
include_graphics("figures/platforms_licenses.png")
```

The majority of tools in the scRNA-tools database have chosen to take an
open-source approach, making their code available under permissive licenses. We
feel this reflects the general underlying sentiment and willingness of the
bioinformatics community to share and build on others work. Variations of the
GNU Public License (GPL) are the most common, covering almost half of tools.
This license allows free use, modification and distribution of source code, but
also has a "copyleft" nature which requires any derivatives to disclose their
source code and use the same license. The MIT license is the second most popular
which also allows use of code for any purpose but without any restrictions on
distribution or licensing. The appropriate license could not be identified for
almost a quarter of tools. This is problematic as fellow developers must assume
that source code cannot be reused, potentially limiting the usefulness of the
methods in those tools. Tool owners are strongly encouraged to clearly display
their license in source code and documentation.

# Categories of scRNA-seq analysis

A typical scRNA-seq analysis can be divided into a number of phases, each of
which consists of various tasks which can be completed using a variety of
analysis tools. In the scRNA-tools database we categorise tools based on the
analysis tasks they perform which can be broadly grouped into four different
phases of analysis [FIGURE]. The data acquisition phase is the first and takes
the raw nucleotide sequences from the sequencing experiment and returns a matrix
describing the expression of each gene (rows) in each cell (columns). This phase
consists of tasks common to bulk RNA-seq experiments, such as alignment to a
reference genome and quantification of expression, but is often extended to
handle Unique Molecular Identifiers (UMIs). Once an expression matrix has been
obtained it is vital to make sure the resulting data is of high enough quality.
In the data cleaning phase (Phase 2) quality control of cells is performed as
well as filtering of uninformative genes. Additional tasks may be performed to
normalise the data or impute missing values and exploratory analysis is often
performed in this phase, such as viewing the datasets in reduced dimensions.

The high-quality expression matrix is the focus of the next phases of analysis.
In Phase 3 cells are assigned, either to discrete groups via clustering or along
a continuous trajectory from one cell type to another. As high-quality reference
datasets become available it will also become feasible to classify cell directly
into different cell types. Once cells have been assigned attention turns to
interpreting what those assignments mean. Identifying interesting genes (Phase
4), such as those that are differentially expressed across groups, marker genes
expressed in a single group or genes that change expression along a trajectory,
is the typical way to do this. Those biological significance of those genes can
then be interpreted to give meaning to the experiment.

```{r analysis-phases, fig.cap = "Phases of the scRNA-seq analysis process. In Phase 1 (data acquisition) raw sequencing reads are converted into a gene by cell expression matrix. For many protocols this requires the alignment of genes to a reference genome and the assignment and de-duplication of Unique Molecular Identifiers (UMIs). The data is then cleaned (Phase 2) to remove low-quality cells and uninformative genes, resulting in a high-quality dataset for further analysis. Missing values can also be imputed during this phase. Phase 3 assigns cells, either in a discrete manner to known (classification) or unknown (clustering) groups or to a position on a continuous trajectory. Interesting genes (eg. differentially expressed, markers, specific patterns of expression) are then identified to explain these groups or trajectories (Phase 4)."}
include_graphics("figures/phase_diagram2.png")
```

While there are other approaches that could be taken to analyse scRNA-seq data
these phases represent the most common path from raw sequencing reads to
biological insight. Descriptions of the categories in the scRNA-tools database are given in [TABLE], along with the associated analysis phases.

```{r cell-assignment, fig.cap = "Strategies for assigning cells. Discrete groups of cells can be formed by clustering cells based on similar expression profiles or classifying them using reference datasets. Alternatively cells can be ordered along a trajectory from one cell type to another.", eval = FALSE}
include_graphics("figures/cell_assignment.png")
```

```{r gene-identification, fig.cap = "Examples of interesting genes to identify. When working with discrete groups tools try to find genes that are differentially expressed across the groups or marker genes that are uniquely expressed in a single group. If cells have been ordered tools instead look for genes that change expression along the trajectory, for example genes that are highly expressed and the start and lowly expressed at the end.", eval = FALSE}
include_graphics("figures/gene_identification.png")
```

```{r categories-table}
cap = "Descriptions of categories for tools in the scRNA-tools database"

cats %>%
    mutate(Category = str_replace_all(Category, "([[:upper:]])", " \\1")) %>%
    arrange(Phase) %>%
    select(Phase, everything()) %>%
    kable(caption = cap)
```

## Trends in scRNA-seq analysis tasks

In this section we look at categories that tools in the scRNA-tools database
have been assigned to and what they tell us about trends in scRNA-seq analysis.
The frequency of the analysis tasks performed reflects the differences between
analysing scRNA-seq and bulk RNA-seq data and the extra opportunities provided
at single-cell resolution. Visualisation is the most commonly performed task and
is important across all stages of analysis for displaying data and results.
Tasks for assigning cells (pseudotime and clustering) are the next most common.
This has been the biggest area of development in single-cell analysis with
clustering tools such as XXXX being used to identify cell types in a sample and
trajectory analysis tools (XXXX) being used to investigate how genes change
across developmental processes.

```{r plot-categories}
phase.cols <- c(brewer.pal(5, "Set1"), "#999999")

cat.counts <- tools %>%
    summarise_at(8:35, sum) %>%
    gather(key = Category, value = Count) %>%
    arrange(-Count) %>%
    mutate(Prop = Count / nrow(tools)) %>%
    left_join(cats, by = "Category") %>%
    mutate(Category = str_replace_all(Category, "([[:upper:]])", " \\1")) %>%
    mutate(Category = str_trim(Category)) %>%
    mutate(Category = factor(Category, levels = Category))

cats.plot <- ggplot(cat.counts,
                    aes(x = Category, weight = Prop, fill = Phase)) +
    geom_bar() +
    scale_y_continuous(labels = scales::percent) +
    scale_fill_manual(values = phase.cols) +
    ylab("Percentage of tools") +
    ggtitle("Analysis categories") +
    theme_cowplot() +
    theme(axis.title.x = element_blank(),
          legend.position = c(0.85, 0.75),
          legend.title = element_text(size = 14),
          legend.text = element_text(size = 12),
          legend.key.size = unit(25, "points"),
          plot.title = element_text(size = 20),
          axis.text = element_text(size = 12),
          axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)
    )

ggsave("figures/categories.png", cats.plot, height = 5)
ggsave("figures/categories.pdf", cats.plot, height = 5)
```

```{r categories, fig.cap = "Categories of tools in the scRNA-tools database. Each tool can be assigned to multiple categories based on the tasks it can complete. Categories associated with multiple analysis phases (visualisation, dimensionality reduction) are the most common, followed by categories associated with the cell assignment phase (pseudotime, clustering)."}
include_graphics("figures/categories.png")
```

* Tools spread across categories
* Common tasks and cell assignment most popular
    * Reflects differences between bulk and single-cell

Dimensionality reduction is also a common task and has applications both in
visualisation (via techniques such as t-SNE) and as a starting point for
analysis. Testing for differential expression is perhaps the most common
analysis performed on bulk RNA-seq datasets and it is also commonly applied by
many scRNA-seq analysis tools, typically to identify genes that are different in
one cluster of cells compared to the rest. However it should be noted that the
DE testing applied by scRNA-seq tools is often not as sophisticated as the
rigorous statistical frameworks of tools developed for bulk RNA-seq such as
edgeR, DESeq2 and limma, often using simple statistical tests such as the
likelihood ratio tests. While methods designed to test DE specifically in
single-cell datasets do exist it is still unclear whether they do or do not
improve on these established methods.

* Decrease in psuedotime tools
* Not really sure what to say about this...

To investigate how the focus of scRNA-seq tool development has changed over time
we have divided the scRNA-tools database into tools added before and after
October 2016. This allowed us to see which analysis tasks are common in more
recently published tools. Many tasks show little change in the proportion of
tools that perform them but there are some areas with a significant change over
time. Specifically, both visualisation and dimensionality reduction are more
commonly performed by recent tools. The quantification category has also seen a
big increase recently as UMI based protocols have become commonly used and tools
designed to handle the extra processing steps required have been developed.
Simulation has always been used to test, develop and validate scRNA-seq tools
but more packages are now including their simulation functions and some tools
have been developed for the specific purpose of generating realistic synthetic
scRNA-seq datasets. More tools are also identifying or making use of
co-regulated gene networks.

```{r plot-categories-time}
cats.time.plot <- tools %>%
    group_by(IsOld) %>%
    mutate(Size = n()) %>%
    group_by(IsOld, Size) %>%
    summarise_at(8:35, sum) %>%
    gather(key = Category, value = Count, -IsOld, -Size) %>%
    mutate(Category = str_replace_all(Category, "([[:upper:]])", " \\1")) %>%
    mutate(Category = str_trim(Category)) %>%
    mutate(Category = factor(Category,
                             levels = rev(levels(cat.counts$Category)))) %>%
    group_by(IsOld) %>%
    mutate(Prop = Count / Size) %>%
    ggplot(aes(x = Category, y = Prop, colour = IsOld)) +
    geom_point(size = 4) +
    scale_color_manual(name = "Date added",
                       labels = c("After 2016-10-01", "Before 2016-10-01"),
                       values = unname(pal)) +
    scale_y_continuous(labels = scales::percent) +
    coord_flip() +
    ylab("Percentage of tools") +
    ggtitle("Analysis categories") +
    theme_cowplot() +
    theme(axis.title.y = element_blank(),
          legend.position = c(0.7, 0.2),
          legend.title = element_text(size = 14),
          legend.text = element_text(size = 12),
          legend.key.size = unit(25, "points"),
          plot.title = element_text(size = 20),
          axis.text = element_text(size = 12)
    )

ggsave("figures/categories_time.png", cats.time.plot, height = 8)
ggsave("figures/categories_time.pdf", cats.time.plot, height = 8)
```

```{r categories-time, fig.cap = "Changes in analysis categories over time, comparing tools added before and after October 2016. There have been significant increases in the percentage of tools associated with visualisation, dimensionality reduction, quantification and simulation. Categories including expression patterns, pseudotime and interactivity have seen relative decreases."}
include_graphics("figures/categories_time.png")
```

Some categories have seen a decrease in the proportion of tools they represent,
most strikingly testing for expression patterns. This is likely related to the
change in pseudotime analysis which is the focus of a slightly lower percentage
of tools added after October 2016.

* Similar information to the last plot but summarised at a higher level

Looking at the tools that can perform at least one task in an analysis phase we
see similar trends. We see increases in the percentage of tools performing tasks
in Phase 1 (quantification), Phase 2 (quality control and filtering), across
multiple phases (visualisation and dimensionality reduction) and alternative
analysis tasks. In contrast the percentage of tools that perform gene
identification tasks (Phase 2) has decreased and the percentage assigning cells
(Phase 3) has remained steady. This may indicate a maturation of the analysis
space as existing tools for performing standard scRNA-seq analyses are deemed
sufficient while there is still room for development in handling data from new
protocols and performing alternative analysis tasks.


```{r plot-date-phases}
phases.plot.date <- tools %>%
    mutate(Phase1 = Quantification,
           Phase2 = QualityControl | Normalisation | Imputation |
               GeneFiltering | CellCycle,
           Phase3 = Clustering | Classification | Pseudotime | RareCells |
               StemCells,
           Phase4 = DifferentialExpression | MarkerGenes | ExpressionPatterns |
               VariableGenes | GeneSets | GeneNetworks,
           Multiple = DimensionalityReduction | Visualisation | Interactive,
           Other = Assembly | Transformation | Modality | AlternativeSplicing |
               Variants | Haplotypes | AlleleSpecific | Simulation) %>%
    select(Date = Added, Phase1, Phase2, Phase3, Phase4, Multiple, Other) %>%
    gather(key = Phase, value = TF, -Date) %>%
    filter(TF == TRUE) %>%
    mutate(Phase = factor(Phase,
                          levels = c("Phase1", "Phase2", "Phase3", "Phase4",
                                     "Multiple", "Other"),
                          labels = c("Phase 1", "Phase 2", "Phase 3", "Phase 4",
                                     "Multiple", "Other"))) %>%
    select(Date, Phase) %>%
    group_by(Date, Phase) %>%
    summarise(PhaseCount = n()) %>%
    group_by(Phase) %>%
    mutate(PhaseTotal = cumsum(PhaseCount)) %>%
    left_join(date.totals, by = "Date") %>%
    mutate(Prop = PhaseTotal / Total) %>%
    ggplot(aes(x = Date, y = Prop, colour = Phase)) +
    geom_line(size = 2) +
    scale_y_continuous(labels = scales::percent) +
    scale_colour_manual(values = phase.cols) +
    xlab("Date") +
    ylab("Percentage of tools") +
    ggtitle("Analysis phases over time") +
    theme_cowplot() +
    theme(plot.title = element_text(size = 20),
          axis.text = element_text(size = 12),
          axis.title.x = element_blank(),
          legend.position = "right"
    )

ggsave("figures/phases_date.png", phases.plot.date, height = 5)
ggsave("figures/phases_date.pdf", phases.plot.date, height = 5)
```

```{r phases-over-time, fig.cap = "Changes in the percentage of tools associated with analysis phases over time. The percentage of tools involved in the data acquisition and data cleaning phases have increased, as have tools completed alternative analysis tasks. The gene identification phase has seen a relative decrease in the number of tools."}
include_graphics("figures/phases_date.png")
```

## Pipelines and toolboxes

* Most tools in few categories
* Later tools have more
* Highlight complete pipelines, tools with many categories

While there are a considerable number of scRNA-seq tools that only perform a
single analysis task, most perform at least two. We find that older tools tend
to specialise more while those that have been added more recently can perform
more tasks. Some tools, such as XXXX are designed to be complete pipelines,
taking raw sequencing reads and producing a report showing the results of a
complete analysis. Others, such as XXXX,  can be thought of as analysis
toolboxes, able to complete a range of complex analyses starting with a gene
expression matrix. Being able to complete multiple tasks using a single tool can
simplify analysis as problems with converting between different data formats can
be avoided, however it is important to remember that it is difficult for a tool
with many functionalities to continue to represent the state of the art in all
of them. Support for common data formats, such as the recently released
SingleCellExperiment object in R, provides another way for developers to allow
easy use of their tools and users to build custom workflows from specialised
tools.

```{r plot-cats-tool}
cats.tools.plot <- tools %>%
    select(Name, 8:35) %>%
    gather(key = Category, value = TF, -Name) %>%
    group_by(Name) %>%
    summarise(Count = sum(TF)) %>%
    ggplot(aes(x = factor(Count, levels = 1:14))) +
    geom_bar(fill = pal["purple"]) +
    scale_x_discrete(drop = FALSE) +
    ggtitle("Number of categories per tool") +
    xlab("Number of categories") +
    ylab("Number of tools") +
    theme_cowplot() +
    theme(plot.title = element_text(size = 20),
          axis.text = element_text(size = 12)
    )

cats.tools.time.plot <- tools %>%
    select(Name, IsOld, 8:35) %>%
    gather(key = Category, value = TF, -Name, -IsOld) %>%
    group_by(Name, IsOld) %>%
    summarise(Cats = sum(TF)) %>%
    ungroup() %>%
    select(-Name) %>%
    group_by(IsOld, Cats) %>%
    mutate(Count = n()) %>%
    unique() %>%
    group_by(IsOld) %>%
    mutate(Prop = Count / sum(Count)) %>%
    ggplot(aes(x = factor(Cats, levels = 1:14), y = Prop, fill = IsOld)) +
    geom_bar(stat = "identity", position = "dodge") +
    scale_fill_manual(name = "Date added",
                      labels = c("After 2016-10-01", "Before 2016-10-01"),
                      values = unname(pal)) +
    scale_y_continuous(labels = scales::percent) +
    ggtitle("Categories per tool") +
    xlab("Number of categories") +
    ylab("Percentage of tools") +
    theme_cowplot() +
    theme(plot.title = element_text(size = 20),
          axis.text = element_text(size = 12),
          legend.position = c(0.7, 0.7)
    )

panel <- plot_grid(cats.tools.plot, cats.tools.time.plot, nrow = 1, ncol = 2,
                   labels = "AUTO")

save_plot("figures/cats_per_tool.png", panel, nrow = 1, ncol = 2,
          base_height = 8)
save_plot("figures/cats_per_tool.pdf", panel, nrow = 1, ncol = 2,
          base_height = 8)
```

```{r cats-per-tool, fig.cap = "The number of categories associated with each tools in the scRNA-tools database. The majority of tools perform few tasks (A). When stratifying by the date tools were added to the database we an increase in the number of categories per tool for tools added at later dates (B)."}
include_graphics("figures/cats_per_tool.png")
```

## Other analyses

* Mention alternative analyses and tools

Some tools perform analyses that lie outside the common tasks performed on
scRNA-seq data. Simulation is one alternative task that has already been
mentioned but there is also a group of tools designed to detect variability
apart from changes in expression. For example alternative splicing, single
nucleotide [?] variants and allele-specific expression. While tools that
complete these tasks are unlikely to ever dominate scRNA-seq analysis it is
likely that we will see an increase in methods for tackling alternative problems
as researchers continue to push the boundaries of what can be observed using
scRNA-seq data.

# Discussion and conclusions

* Where to from here?
    * More tools
    * Improvements and updates
    * New tasks
    * New datasets
        * Integration
    * New technologies
    * Reference databases

As scRNA-seq analysis matures we can expect some existing tools to continue to
add functionality while others will cease to be updated and maintained.

# Methods

# Declarations

# Ethics

Not applicable.

# Availability of data and materials

# Competing interests

The authors declare no competing interests.

# Funding

Luke Zappia is supported by an Australian Government Research Training Program
(RTP) Scholarship. Alicia Oshlack is supported through a National Health and
Medical Research Council Career Development Fellowship APP1126157. MCRI is
supported by the Victorian Government's Operational Infrastructure Support
Program.

# Authors' contributions

# Acknowledgements

* Sean Davis
* Licenses...
* Breon

# References

# Additional files
