---
title: "Exploring the single-cell RNA-seq analysis landscape"
author:
- Luke Zappia
- Belinda Phipson
- Alicia Oshlack
date: "`r Sys.Date()`"
output:
    bookdown::word_document2:
        reference_docx: style/style.docx
        pandoc_args: [ "--csl", "style/nature.csl" ]
bibliography: style/references.bib
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(autodep        = TRUE,
                      cache          = FALSE,
                      cache.path     = "cache/",
                      cache.comments = TRUE,
                      echo           = FALSE,
                      error          = FALSE,
                      fig.path       = "figures/",
                      fig.width      = 10,
                      fig.height     = 8,
                      dev            = c('png', 'pdf'),
                      message        = FALSE,
                      warning        = FALSE)
```

```{r libaries}
# Plotting
library("cowplot")
library("RColorBrewer")

# Presentation
library("knitr")

# Strings
library("stringr")

# JSON
library("jsonlite")

# Tidyverse
library("tidyverse")
```

```{r palette}
pal <- c(pink   = "#EC008C",  # Pink
         blue   = "#00ADEF",  # Blue
         green  = "#8DC63F",  # Green
         teal   = "#00B7C6",  # Teal
         orange = "#F47920",  # Orange
         purple = "#7A52C7")  # Purple
```

```{r load}
tools <- read_csv("data/single_cell_software.csv",
                  col_types = cols(
                      .default = col_logical(),
                      Name = col_character(),
                      Platform = col_character(),
                      DOI = col_character(),
                      PubDate = col_character(),
                      Code = col_character(),
                      Description = col_character(),
                      License = col_character(),
                      Added = col_date(format = ""),
                      Updated = col_date(format = "")
                  )) %>%
    mutate(IsOld = Added < "2016-10-01")

cats <- read_json("data/descriptions.json", simplifyVector = TRUE) %>%
    mutate(Phase = factor(Phase,
                          levels = c("Phase 1", "Phase 2", "Phase 3", "Phase 4",
                                     "Multiple", "Other")))
```

# Abstract

As single-cell RNA-sequencing (scRNA-seq) datasets have become more widespread
the number of tools designed to anlyse these data has dramatically increased. We
have been cataloguing and curating new analysis tools, as they become available,
in the scRNA-tools database (www.scRNA-tools.org). Our database collects a range
of information on each scRNA-seq analysis tool and categorises tools according
to the analysis tasks they perform. Exploration of this database gives insights
into the areas of rapid development of analysis methods for scRNA-seq data.
Specifically we see that many tools are developed to perform tasks specific to
scRNA-seq analysis, particularly and ordering of cells. We also find that the
scRNA-seq community takes an open-source approach, with most tools making their
code available under open-source licenses and preprints being extensively uses
as a means to describe methods. The scRNA-seq tools database provides a resource
for researchers approaching analysis of an scRNA-seq dataset and a record of the
field over time.

# Keywords

# Introduction

Single-cell RNA-sequencing (scRNA-seq) has rapidly gained traction as an
effective tool for interrogating the transcriptome at the resolution of
individual cells. Since the first protocols were published in 2009
[@Tang2009-jd] the number of cells profiled in individual scRNA-seq experiments
has increased exponentially, outstripping Mooreâ€™s Law [@Svensson2017-es]. This
new kind of transcriptomic data brings a demand for new analysis methods. Not
only does the scale of scRNA- seq datasets vastly outstrip bulk experiments but
there are also a variety of challenges unique to the single-cell context
[@Stegle2015-mc]. Specifically, scRNA-seq data is extremely sparse (there is no
expression measured for many genes in most cells), it contains large technical
artefacts and the scientific questions of interest are often different to those
asked of bulk RNA-seq datasets. For example many bulk RNA-seq datasets are
generated to detect differentially expressed genes through a designed experiment
while many scRNA-seq experiments aim to identify or classify cell types.

The bioinformatics community has embraced this new type of data with the
production of a plethora of methods for the analysis of scRNA-seq data. As such,
keeping up with the current state of scRNA-seq analysis is now a significant
challenge as the field is presented with a huge number of choices for
approaching an analysis. Since September 2016 we have collated and categorised
scRNA-seq analysis tools as they have become available. This database is being
continually updated and is publicly available at www.scRNA-tools.org. In order
to help researchers navigate the analysis jungle we discuss the stages of
scRNA-seq analysis and their relationship to tools and categories in the
scRNA-tools database. Through the analysis of this database we show trends in
not only the analysis applications their methods address but how they are
published, licensed and the platforms they use. Based on this database we gain
insight into the state of analysis tools in this rapidly developing field.

# Overview of the scRNA-tools database

The scRNA-tools database contains information on software tools specifically
designed for the analysis of scRNA-seq data. For a tool to be eligible to be
included in the database it must be available for download and public use. This
can be from a software package repository (such as Bioconductor [@Huber2015-lz],
CRAN or PyPI), a code sharing website such as GitHub or directly from a private
website. Various details of the tools are recorded such as the programming
language or platform they use, details of any related publication, links to the
source code and the associated software license. Tools are also categorised
according to the analysis tasks they are able to perform. Most tools are added
after a preprint or publication becomes available but some have been added after
being mentioned on social media or in similar collections such as Sean Davis'
awesome-single-cell page (https://github.com/seandavi/awesome-single-cell).

Information about tools is displayed on a publicly available website at
www.scRNA-tools.org. This website provides a profile for each tool, with links
to publications and code repositories, as well as an index by analysis category.
We also provide an interactive table that allows users to filter and sort tools
to find those most relevant to their needs. Anyone wishing to contribute to the
database can do so by submitting an issue to the GitHub page
(https://github.com/Oshlack/scRNA-tools).

When the database was first constructed there were 70 scRNA-seq analysis tools
available, representing work in the field during the three years from the first
published tool in November 2013 (SAMstrt [@Katayama2013-pq]) up to October 2016.
In the year since then over 60 more tools have been added (Figure
\@ref(fig:tools-over-time)). The doubling of the number of available tools in
such a short time demonstrates the blossoming interest in scRNA-seq and its
maturation from a technique requiring custom-built equipment with specialised
protocols to a commercially available product.

```{r plot-tools-over-time}
date.totals <- tools %>%
    select(Date = Added) %>%
    group_by(Date = as.Date(Date)) %>%
    summarise(Count = n()) %>%
    complete(Date = full_seq(Date, 1), fill = list(Count = 0)) %>%
    mutate(Total = cumsum(Count))

totals.plot <- ggplot(date.totals, aes(x = Date, y = Total)) +
    geom_line(size = 2, colour = pal["purple"]) +
    xlab("Date") +
    ylab("Number of tools") +
    scale_x_date(breaks = scales::pretty_breaks(10)) +
    ggtitle("Increase in tools over time") +
    theme_cowplot() +
    theme(plot.title = element_text(size = 20),
          axis.title.x = element_blank(),
          axis.text = element_text(size = 12),
          axis.text.x = element_text(angle = 90, vjust = 0.5)
    )

ggsave("figures/tools_over_time.png", totals.plot, height = 6)
ggsave("figures/tools_over_time.pdf", totals.plot, height = 6)
```

```{r tools-over-time, fig.cap = "Number of tools in the scRNA-tools database over time. Since the scRNA-seq tools database was started in September 2016 more than 60 new tools have been released."}
include_graphics("figures/tools_over_time.png")
```

## Publication status

Most tools have been added to the scRNA-tools database after coming to our
attention in a paper describing their method and use. Of all the tools in the
database about half have been published in peer-reviewed journals and another
third are described in preprint articles, typically on the bioRxiv preprint
server (Figure \@ref(fig:publication-status)A). When the tools are split into
two groups based on the date they were added to the scRNA-tools database we can
see that the majority of older tools have been published while more recent tools
are more likely to only be available as preprints (Figure
\@ref(fig:publication-status)B). This is a good demonstration of the delay
imposed by the traditional publication process. By publishing preprints and
releasing software available via repositories such as GitHub scRNA-seq tool
developers make their methods available to the community much earlier, allowing
them to be used for analysis and their methods improved prior to formal
publication.

```{r plot-publication-status}
pub.plot <- tools %>%
    select(PubDate) %>%
    mutate(IsPre = PubDate == "PREPRINT") %>%
    mutate(IsPub = !is.na(PubDate) & !IsPre) %>%
    mutate(IsNot = is.na(PubDate)) %>%
    summarise(NotPublished = sum(IsNot),
              Published = sum(IsPub, na.rm = TRUE),
              Preprint = sum(IsPre, na.rm = TRUE)) %>%
    gather(key = Type, value = Count) %>%
    mutate(Type = factor(Type,
                         levels = c("Published", "Preprint", "NotPublished"),
                         labels = c("Published", "Preprint", "Not Published")),
           Cumulative = cumsum(Count),
           Midpoint = Cumulative - (Count / 2),
           Label = paste0(Type, "\n",
                          round(Count / sum(Count) * 100, 1), "%")) %>%
    ggplot(aes(x = 1, weight = Count, fill = Type)) +
    geom_bar(width = 1, position = "stack") +
    coord_polar(theta = "y") +
    geom_text(aes(x = 1.8, y = Midpoint, label = Label, colour = Type),
              size = 6) +
    scale_fill_manual(values = unname(pal)) +
    scale_colour_manual(values = unname(pal)) +
    ggtitle("Publication status") +
    theme_nothing() +
    theme(plot.title = element_text(size = 30),
          legend.position = "none"
    )

pub.plot.time <- tools %>%
    group_by(IsOld) %>%
    select(PubDate, IsOld) %>%
    mutate(IsPre = PubDate == "PREPRINT") %>%
    mutate(IsPub = !is.na(PubDate) & !IsPre) %>%
    mutate(IsNot = is.na(PubDate)) %>%
    summarise(NotPublished = sum(IsNot),
              Published = sum(IsPub, na.rm = TRUE),
              Preprint = sum(IsPre, na.rm = TRUE)) %>%
    gather(key = Type, value = Count, -IsOld) %>%
    group_by(IsOld) %>%
    mutate(Prop = Count / sum(Count)) %>%
    mutate(Type = factor(Type,
                         levels = c("Preprint", "Published", "NotPublished"),
                         labels = c("Preprint", "Published", "Not Published"))) %>%
    ggplot(aes(x = Type, y = Prop, fill = IsOld)) +
    geom_bar(stat = "identity", position = "dodge") +
    scale_fill_manual(name = "Date added",
                      labels = c("After 2016-10-01", "Before 2016-10-01"),
                      values = unname(pal)) +
    scale_y_continuous(labels = scales::percent) +
    ggtitle("Publication status over time") +
    ylab("Percentage of tools") +
    theme_cowplot() +
    theme(legend.position = c(0.75, 0.85),
          legend.title = element_text(size = 14),
          legend.text = element_text(size = 12),
          legend.key.size = unit(25, "points"),
          plot.title = element_text(size = 20),
          axis.text = element_text(size = 12),
          axis.title.x = element_blank()
    )

panel <- plot_grid(pub.plot, pub.plot.time, nrow = 1, ncol = 2,
                   labels = "AUTO")

save_plot("figures/publication_status.png", panel, nrow = 1, ncol = 2,
          base_height = 8)
save_plot("figures/publication_status.pdf", panel, nrow = 1, ncol = 2,
          base_height = 8)
```

```{r publication-status, fig.cap = "Publication status of tools in the scRNA-tools database. Over half of the tools in the full database have been published in peer-revewied journals while another third are available as preprints (A). When stratified by the date tools were added to the database we see that the majority of tools added before October 2016 are published, while the majority of newer tools are available as preprints (B). The number of unpublished tools has stayed consistent at around 10 percent."}
include_graphics("figures/publication_status.png")
```

## Open-source

```{r plot-licenses}
license.plot <- tools %>%
    select(License) %>%
    mutate(IsGPL = str_detect(License, "GPL"),
           IsBSD = str_detect(License, "BSD"),
           IsMIT = str_detect(License, "MIT"),
           IsApache = str_detect(License, "Apache"),
           IsArtistic = str_detect(License, "Artistic"),
           IsUnknown = is.na(License),
           IsOther = !(IsGPL | IsBSD | IsMIT | IsApache | IsArtistic |
                           IsUnknown)) %>%
    summarise(Apache = sum(IsApache, na.rm = TRUE),
              Artistic = sum(IsArtistic, na.rm = TRUE),
              BSD = sum(IsBSD, na.rm = TRUE),
              GPL = sum(IsGPL, na.rm = TRUE),
              MIT = sum(IsMIT, na.rm = TRUE),
              Other = sum(IsOther),
              Unknown = sum(IsUnknown)) %>%
    gather(key = License, value = Count) %>%
    mutate(License = factor(License,
                            levels = c("Apache", "Artistic", "BSD", "GPL",
                                       "MIT", "Other", "Unknown")),
           Cumulative = cumsum(Count),
           Midpoint = Cumulative - (Count / 2),
           Label = paste0(License, "\n",
                          round(Count / sum(Count) * 100, 1), "%")) %>%
    ggplot(aes(x = 1, weight = Count, fill = License)) +
    geom_bar(width = 1, position = "stack") +
    coord_polar(theta = "y") +
    geom_text(aes(x = 1.75, y = nrow(tools) - Midpoint, label = Label,
                  colour = License), size = 6) +
    scale_fill_manual(values = c(unname(pal), "#999999")) +
    scale_colour_manual(values = c(unname(pal), "#999999")) +
    ggtitle("Software licenses") +
    theme_nothing() +
    theme(plot.title = element_text(size = 30),
          legend.position = "none"
    )
```

```{r plot-platforms}
platforms.plot <- tools %>%
    select(Platform) %>%
    mutate(IsR = str_detect(Platform, "R"),
           IsPython = str_detect(Platform, "Python"),
           IsMATLAB = str_detect(Platform, "MATLAB"),
           IsCPP = str_detect(Platform, "C++"),
           IsOther = !(IsR | IsPython | IsMATLAB | IsCPP)) %>%
    summarise(R = sum(IsR),
              Python = sum(IsPython),
              MATLAB = sum(IsMATLAB),
              CPP = sum(IsCPP),
              Other = sum(IsOther)) %>%
    gather(key = Platform, value = Count) %>%
    mutate(Platform = factor(Platform,
                             levels = c("R", "Python", "MATLAB", "CPP",
                                        "Other"),
                             labels = c("R", "Python", "MATLAB", "C++",
                                        "Other"))) %>%
    ggplot(aes(x = Platform, weight = Count, fill = Platform)) +
    geom_bar(aes(y = (..count..) / sum(..count..))) +
    scale_y_continuous(labels = scales::percent) +
    scale_fill_manual(values = unname(pal)) +
    ylab("Percentage of tools") +
    ggtitle("Platforms used by analysis tools") +
    theme_cowplot() +
    theme(axis.title.x = element_blank(),
          legend.position = "none",
          plot.title = element_text(size = 20),
          axis.text = element_text(size = 12)
    )
```

```{r plot-platforms-licenses}
panel <- plot_grid(platforms.plot, license.plot, nrow = 1, ncol = 2,
                   labels = "AUTO")

save_plot("figures/platforms_licenses.png", panel, nrow = 1, ncol = 2,
          base_height = 8)
save_plot("figures/platforms_licenses.pdf", panel, nrow = 1, ncol = 2,
          base_height = 8)
```

Developers of scRNA-seq analysis tools have choices to make about what platforms
they use to create their tools, how they make them available to the community
and whether they share the source code. We find that the most commonly used
platform for creating scRNA-seq analysis tools is the R statistical programming
language, with many tools being available through the Bioconductor or CRAN
repositories (Figure \@ref(fig:platforms-licenses)A). Python is the second most
popular language, followed by MATLAB, a commercially available product, and the
lower-level C++. The use of R and Python is consistent with their popularity
across a range of data science fields. In particular the use of R reflects its
history as the language of choice for the analysis of bulk RNA-seq datasets and
a range of other biological data types.

```{r platforms-licenses, fig.cap = "The open-source nature of scRNA-seq analysis tools. The majority of tools are available using either the R or Python programming languages (A). Most tools are released under a standard open-source software license, with variants of the GNU Public License (GPL) being the most common. However licenses could not be found for a large proportion of tools."}
include_graphics("figures/platforms_licenses.png")
```

The majority of tools in the scRNA-tools database have chosen to take an
open-source approach, making their code available under permissive licenses
(Figure \@ref(fig:platforms-licenses)B). We feel this reflects the general
underlying sentiment and willingness of the bioinformatics community to share
and build on others work. Variations of the GNU Public License (GPL) are the
most common, covering almost half of tools. This license allows free use,
modification and distribution of source code, but also has a "copyleft" nature
which requires any derivatives to disclose their source code and use the same
license. The MIT license is the second most popular which also allows use of
code for any purpose but without any restrictions on distribution or licensing.
The appropriate license could not be identified for almost a fifth of tools.
This is problematic as fellow developers must assume that source code cannot be
reused, potentially limiting the usefulness of the methods in those tools. Tool
owners are strongly encouraged to clearly display their license in source code
and documentation.

# Categories of scRNA-seq analysis

A typical scRNA-seq analysis can be divided into a number of phases, each of
which consists of various tasks which can be completed using a variety of
analysis tools. In the scRNA-tools database we categorise tools based on the
analysis tasks they perform which can be broadly grouped into four different
phases of analysis (Figure \@ref(fig:analysis-phases)). The data acquisition
phase is the first and takes the raw nucleotide sequences from the sequencing
experiment and returns a matrix describing the expression of each gene (rows) in
each cell (columns). This phase consists of tasks common to bulk RNA-seq
experiments, such as alignment to a reference genome and quantification of
expression, but is often extended to handle Unique Molecular Identifiers (UMIs).
Once an expression matrix has been obtained it is vital to make sure the
resulting data is of high enough quality. In the data cleaning phase (Phase 2)
quality control of cells is performed as well as filtering of uninformative
genes. Additional tasks may be performed to normalise the data or impute missing
values and exploratory analysis is often performed in this phase, such as
viewing the datasets in reduced dimensions.

The high-quality expression matrix is the focus of the next phases of analysis.
In Phase 3 cells are assigned, either to discrete groups via clustering or along
a continuous trajectory from one cell type to another. As high-quality reference
datasets become available it will also become feasible to classify cell directly
into different cell types. Once cells have been assigned attention turns to
interpreting what those assignments mean. Identifying interesting genes (Phase
4), such as those that are differentially expressed across groups, marker genes
expressed in a single group or genes that change expression along a trajectory,
is the typical way to do this. The biological significance of those genes can
then be interpreted to give meaning to the experiment.

```{r analysis-phases, fig.cap = "Phases of the scRNA-seq analysis process. In Phase 1 (data acquisition) raw sequencing reads are converted into a gene by cell expression matrix. For many protocols this requires the alignment of genes to a reference genome and the assignment and de-duplication of Unique Molecular Identifiers (UMIs). The data is then cleaned (Phase 2) to remove low-quality cells and uninformative genes, resulting in a high-quality dataset for further analysis. The data can also be normalised and missing values imputed during this phase. Phase 3 assigns cells, either in a discrete manner to known (classification) or unknown (clustering) groups or to a position on a continuous trajectory. Interesting genes (eg. differentially expressed, markers, specific patterns of expression) are then identified to explain these groups or trajectories (Phase 4)."}
include_graphics("figures/phase_diagram2.png")
```

While there are other approaches that could be taken to analyse scRNA-seq data
these phases represent the most common path from raw sequencing reads to
biological insight. Descriptions of the categories in the scRNA-tools database
are given in Table \@ref(tab:categories-table), along with the associated
analysis phases.

```{r categories-table}
cap <- "Descriptions of categories for tools in the scRNA-tools database"

cats %>%
    mutate(Category = str_replace_all(Category, "([[:upper:]])", " \\1")) %>%
    mutate(Category = str_trim(Category)) %>%
    mutate(Category = ifelse(Category == "U M Is", "UMIs", Category)) %>%
    arrange(Phase) %>%
    select(Phase, everything()) %>%
    kable(caption = cap)
```

## Trends in scRNA-seq analysis tasks

Each of the tools in the database is assigned to one or more analysis
categories. We investigated these categories in further detail to give insight
into the trends in scRNA-seq analysis. Figure \@ref(fig:categories) shows the
frequency of tools performing each of the analysis tasks. Visualisation is the
most commonly included task and is important across all stages of analysis for
exploring and displaying data and results. Tasks for assigning cells (ordering
and clustering) are the next most common. This has been the biggest area of
development in single-cell analysis with clustering tools such as Seurat
[@Satija2015-or], SC3 [@Kiselev2017-an] and BackSPIN [@Zeisel2015-rd] being used
to identify cell types in a sample and trajectory analysis tools (Monocle
[@Trapnell2014-he], Wishbone [@Setty2016-yh], DPT [@Haghverdi2016-zt]) being
used to investigate how genes change across developmental processes. These areas
of analysis reflect the extra opportunities provided by single-cell resolution
over bulk RNA-seq experiments.

```{r plot-categories}
phase.cols <- c(brewer.pal(5, "Set1"), "#999999")

cat.counts <- tools %>%
    summarise_at(8:38, sum) %>%
    gather(key = Category, value = Count) %>%
    arrange(-Count) %>%
    mutate(Prop = Count / nrow(tools)) %>%
    left_join(cats, by = "Category") %>%
    mutate(Category = str_replace_all(Category, "([[:upper:]])", " \\1")) %>%
    mutate(Category = str_trim(Category)) %>%
    mutate(Category = ifelse(Category == "U M Is", "UMIs", Category)) %>%
    mutate(Category = factor(Category, levels = Category))

cats.plot <- ggplot(cat.counts,
                    aes(x = Category, weight = Prop, fill = Phase)) +
    geom_bar() +
    scale_y_continuous(labels = scales::percent) +
    scale_fill_manual(values = phase.cols) +
    ylab("Percentage of tools") +
    ggtitle("Analysis categories") +
    theme_cowplot() +
    theme(axis.title.x = element_blank(),
          legend.position = c(0.85, 0.70),
          legend.title = element_text(size = 14),
          legend.text = element_text(size = 12),
          legend.key.size = unit(25, "points"),
          plot.title = element_text(size = 20),
          axis.text = element_text(size = 12),
          axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)
    )

ggsave("figures/categories.png", cats.plot, height = 6)
ggsave("figures/categories.pdf", cats.plot, height = 6)
```

```{r categories, fig.cap = "Categories of tools in the scRNA-tools database. Each tool can be assigned to multiple categories based on the tasks it can complete. Categories associated with multiple analysis phases (visualisation, dimensionality reduction) are among the most common, as are categories associated with the cell assignment phase (ordering, clustering)."}
include_graphics("figures/categories.png")
```

Dimensionality reduction is also a common task and has applications in
visualisation (via techniques such as t-SNE [@Maaten2008-ne]), quality control
and as a starting point for analysis. Testing for differential expression (DE)
is perhaps the most common analysis performed on bulk RNA-seq datasets and it is
also commonly applied by many scRNA-seq analysis tools, typically to identify
genes that are different in one cluster of cells compared to the rest. However
it should be noted that the DE testing applied by scRNA-seq tools is often not
as sophisticated as the rigorous statistical frameworks of tools developed for
bulk RNA-seq such as edgeR [@Robinson2010-pt, @McCarthy2012-gc], DESeq2
[@Love2014-tw] and limma [@Ritchie2015-te], often using simple statistical tests
such as the likelihood ratio test. While methods designed to test DE
specifically in single-cell datasets do exist (SCDE [@Kharchenko2014-tb], scDD
[@Korthauer2016-yq]) it is still unclear whether they improve on methods that
have been established for bulk data [@Soneson2017-gh, @Jaakkola2016-ko,
@Miao2016-eo, @Dal_Molin2017-pl].

To investigate how the focus of scRNA-seq tool development has changed over time
we again divided the scRNA-tools database into tools added before and after
October 2016. This allowed us to see which analysis tasks are more common in
recently published tools. We looked at the percentage of tools in each time
period that performed tasks in the different analysis categories. Some
categories show little change in the proportion of tools that perform them but
there are other areas with a more significant change over time. Specifically,
both visualisation and dimensionality reduction are more commonly addressed by
recent tools. The UMIs category has also seen a big increase recently as UMI
based protocols have become commonly used and tools designed to handle the extra
processing steps required have been developed (UMI-tools [@Smith2017-nz], umis
[@Svensson2017-zo], zUMIs [@Parekh2017-hq]). Simulation has always been used to
test, develop and validate scRNA-seq tools but more packages are now including
their simulation functions and some tools have been developed for the specific
purpose of generating realistic synthetic scRNA-seq datasets (powsimR
[@Vieth2017-il], Splatter [@Zappia2017-sg]). Classification of cells in known
groups has also increased as reference datasets become available and more tools
are identifying or making use of co-regulated gene networks.

```{r plot-categories-time}
cats.time.points <- tools %>%
    group_by(IsOld) %>%
    mutate(Size = n()) %>%
    group_by(IsOld, Size) %>%
    summarise_at(8:38, sum) %>%
    gather(key = Category, value = Count, -IsOld, -Size) %>%
    mutate(Category = str_replace_all(Category, "([[:upper:]])", " \\1")) %>%
    mutate(Category = str_trim(Category)) %>%
    mutate(Category = ifelse(Category == "U M Is", "UMIs", Category)) %>%
    mutate(Category = factor(Category,
                             levels = rev(levels(cat.counts$Category)))) %>%
    group_by(IsOld) %>%
    mutate(Prop = Count / Size)

cats.time.changes <- cats.time.points %>%
    select(Category, IsOld, Prop) %>%
    spread(IsOld, Prop) %>%
    rename(Old = `TRUE`, New = `FALSE`) %>%
    mutate(Decrease = New < Old)

cats.time.plot <- ggplot(cats.time.points, aes(x = Category)) +
    geom_linerange(data = cats.time.changes,
                   aes(ymin = Old, ymax = New, colour = Decrease),
                   size = 4.5, alpha = 0.3) +
    geom_point(aes(y = Prop, colour = IsOld), size = 4) +
    scale_color_manual(name = "Date added",
                       labels = c("After 2016-10-01", "Before 2016-10-01"),
                       values = unname(pal)) +
    scale_y_continuous(labels = scales::percent) +
    coord_flip() +
    ylab("Percentage of tools") +
    ggtitle("Analysis categories") +
    theme_cowplot() +
    theme(axis.title.y = element_blank(),
          legend.position = c(0.7, 0.2),
          legend.title = element_text(size = 14),
          legend.text = element_text(size = 12),
          legend.key.size = unit(25, "points"),
          plot.title = element_text(size = 20),
          axis.text = element_text(size = 12),
          plot.margin = unit(c(1, 1, 0.5, 0.5), "lines")
    )

ggsave("figures/categories_time.png", cats.time.plot, height = 8)
ggsave("figures/categories_time.pdf", cats.time.plot, height = 8)
```

```{r categories-time, fig.cap = "Changes in analysis categories over time, comparing tools added before and after October 2016. There have been significant increases in the percentage of tools associated with visualisation, dimensionality reduction, quantification and simulation. Categories including expression patterns, pseudotime and interactivity have seen relative decreases."}
include_graphics("figures/categories_time.png")
```

Some categories have seen a decrease in the proportion of tools they represent,
most strikingly testing for changes expression patterns along a trajectory. This
is likely related to the change in cell ordering analysis which is the focus of
a lower percentage  of tools added after October 2016. The ordering of cells
along a trajectory was one of the first developments in scRNA-seq analysis and a
decrease in the development of these tools could indicate that researchers have
moved on to other techniques or that use has converged on a set of mature tools.

By grouping categories based on their associated analysis phases we see similar
trends. We see increases in the percentage of tools performing tasks in Phase 1
(quantification), Phase 2 (quality control and filtering), across multiple
phases (visualisation and dimensionality reduction) and alternative analysis
tasks. In contrast the percentage of tools that perform gene identification
tasks (Phase 2) has decreased and the percentage assigning cells (Phase 3) has
remained steady. This too may indicate a maturation of the analysis space as
existing tools for performing standard scRNA-seq analyses are deemed sufficient
while there is still room for development in handling data from new protocols
and performing alternative analysis tasks.

```{r plot-date-phases}
phases.plot.date <- tools %>%
    mutate(Phase1 = Alignment | Assembly | Quantification | UMIs,
           Phase2 = CellCycle | GeneFiltering | Imputation | Normalisation |
               QualityControl,
           Phase3 = Classification | Clustering | Ordering | RareCells | StemCells,
           Phase4 = DifferentialExpression | ExpressionPatterns | GeneNetworks |
               GeneSets | MarkerGenes | VariableGenes,
           Multiple = DimensionalityReduction | Visualisation | Interactive,
           Other = AlleleSpecific | AlternativeSplicing | Haplotypes | Immune |
               Modality | Simulation | Transformation | Variants) %>%
    select(Date = Added, Phase1, Phase2, Phase3, Phase4, Multiple, Other) %>%
    gather(key = Phase, value = TF, -Date) %>%
    filter(TF == TRUE) %>%
    mutate(Phase = factor(Phase,
                          levels = c("Phase1", "Phase2", "Phase3", "Phase4",
                                     "Multiple", "Other"),
                          labels = c("Phase 1", "Phase 2", "Phase 3", "Phase 4",
                                     "Multiple", "Other"))) %>%
    select(Date, Phase) %>%
    group_by(Date, Phase) %>%
    summarise(PhaseCount = n()) %>%
    ungroup() %>%
    complete(Date = full_seq(Date, 1), Phase, fill = list(PhaseCount = 0)) %>%
    group_by(Phase) %>%
    mutate(PhaseTotal = cumsum(PhaseCount)) %>%
    left_join(date.totals, by = "Date") %>%
    mutate(Prop = PhaseTotal / Total) %>%
    ggplot(aes(x = Date, y = Prop, colour = Phase)) +
    geom_line(size = 2) +
    scale_x_date(breaks = scales::pretty_breaks(10)) +
    scale_y_continuous(labels = scales::percent) +
    scale_colour_manual(values = phase.cols) +
    xlab("Date") +
    ylab("Percentage of tools") +
    ggtitle("Analysis phases over time") +
    theme_cowplot() +
    theme(plot.title = element_text(size = 20),
          axis.text = element_text(size = 12),
          axis.text.x = element_text(angle = 90, vjust = 0.5),
          axis.title.x = element_blank(),
          legend.position = "right"
    )

ggsave("figures/phases_date.png", phases.plot.date, height = 8)
ggsave("figures/phases_date.pdf", phases.plot.date, height = 8)
```

```{r phases-over-time, fig.cap = "Changes in the percentage of tools associated with analysis phases over time. The percentage of tools involved in the data acquisition and data cleaning phases have increased, as have tools designed for alternative analysis tasks. The gene identification phase has seen a relative decrease in the number of tools."}
include_graphics("figures/phases_date.png")
```

## Pipelines and toolboxes

While there are a considerable number of scRNA-seq tools that only perform a
single analysis task, many perform at least two. Some tools (dropEst
[@Petukhov2017-lm], DrSeq2 [@Zhao2017-ju], scPipe [@Tian2017-zl]) are
preprocessing pipelines, taking raw sequencing reads and producing an expression
matrix. Others, such as Scanpy [@Alexander_Wolf2017-hp], SCell [@Diaz2016-fz],
Seurat, Monocle and scater [@McCarthy2017-ql] can be thought of as analysis
toolboxes, able to complete a range of complex analyses starting with a gene
expression matrix. Most of the tools that complete many tasks are more recent.
Being able to complete multiple tasks using a single tool can simplify analysis
as problems with converting between different data formats can be avoided,
however it is important to remember that it is difficult for a tool with many
functionalities to continue to represent the state of the art in all of them.
Support for common data formats, such as the recently released
SingleCellExperiment object in R [@Lun2017-jj], provides another way for
developers to allow easy use of their tools and users to build custom workflows
from specialised tools.

```{r plot-cats-tool}
cats.tools.plot <- tools %>%
    select(Name, 8:38) %>%
    gather(key = Category, value = TF, -Name) %>%
    group_by(Name) %>%
    summarise(Count = sum(TF)) %>%
    ggplot(aes(x = factor(Count, levels = 1:14))) +
    geom_bar(fill = pal["purple"]) +
    scale_x_discrete(drop = FALSE) +
    ggtitle("Number of categories per tool") +
    xlab("Number of categories") +
    ylab("Number of tools") +
    theme_cowplot() +
    theme(plot.title = element_text(size = 20),
          axis.text = element_text(size = 12)
    )

cats.tools.time.plot <- tools %>%
    select(Name, IsOld, 8:38) %>%
    gather(key = Category, value = TF, -Name, -IsOld) %>%
    group_by(Name, IsOld) %>%
    summarise(Cats = sum(TF)) %>%
    ungroup() %>%
    select(-Name) %>%
    group_by(IsOld, Cats) %>%
    mutate(Count = n()) %>%
    unique() %>%
    group_by(IsOld) %>%
    mutate(Prop = Count / sum(Count)) %>%
    ungroup() %>%
    complete(IsOld, Cats, fill = list(Count = 0, Prop = 0)) %>%
    ggplot(aes(x = factor(Cats, levels = 1:14), y = Prop, fill = IsOld)) +
    geom_bar(stat = "identity", position = "dodge") +
    scale_fill_manual(name = "Date added",
                      labels = c("After 2016-10-01", "Before 2016-10-01"),
                      values = unname(pal)) +
    scale_x_discrete(drop = FALSE) +
    scale_y_continuous(labels = scales::percent) +
    ggtitle("Categories per tool by date added") +
    xlab("Number of categories") +
    ylab("Percentage of tools") +
    theme_cowplot() +
    theme(plot.title = element_text(size = 20),
          axis.text = element_text(size = 12),
          legend.position = c(0.7, 0.7)
    )

panel <- plot_grid(cats.tools.plot, cats.tools.time.plot, nrow = 1, ncol = 2,
                   labels = "AUTO")

save_plot("figures/cats_per_tool.png", panel, nrow = 1, ncol = 2,
          base_height = 8)
save_plot("figures/cats_per_tool.pdf", panel, nrow = 1, ncol = 2,
          base_height = 8)
```

```{r cats-per-tool, fig.cap = "The number of categories associated with each tools in the scRNA-tools database. The majority of tools perform few tasks (A). Most tools that complete many tasks have are relatively recent. (B)."}
include_graphics("figures/cats_per_tool.png")
```

## Alternative analyses

Some tools perform analyses that lie outside the common tasks performed on
scRNA-seq data described above. Simulation is one alternative task that has
already been mentioned but there is also a group of tools designed to detect
biological signals in scRNA-seq data apart from changes in expression. For
example alternative splicing (BRIE [@Huang2017-lp], Outrigger [@Song2017-ym],
SingleSplice [@Welch2016-dv]), single nucleotide variants (SSrGE
[@Poirion2016-ld]) and allele-specific expression (SCALE [@Jiang2017-sb]).
Reconstruction of immune cell receptors is another area that has received
considerable attention from tools such as BASIC [@Canzar2017-jb], TraCeR
[@Stubbington2016-xw] and TRAPeS [@Afik2017-qw]. While tools that complete these
tasks are unlikely to ever dominate scRNA-seq analysis it is likely that we will
see an increase in methods for tackling specialised analyses as researchers
continue to push the boundaries of what can be observed using scRNA-seq data.

# Discussion and conclusions

Over the last year we have seen the number of number of software tools for
analysing single-cell RNA-seq data double, with more than 130 anlysis tools now
available. As new tools have become available we have curated and catalogued
them in the scRNA-tools database where we record the analysis tasks that they
can complete, along with additional information such as any associated
publications. By analysing this database we have found that tool developers have
focused much of their efforts on methods for handling new problems specific to
scRNA-seq data, in particular clustering cells into groups or ordering them
along a trajectory. We have also seen that the scRNA-seq community is generally
open and willing to share with most methods being described in preprint prior to
publication and the majority of software being released under permissive
open-source licenses.

The next few years promise to continue to produce significant new developments
in scRNA-seq analysis. New tools will continue to be produced, becoming
increasingly sophisticated and aiming to address more of the questions made
possible by scRNA-seq data. We anticipate that some existing tools will continue
to improve and expand their functionality while others will cease to be updated
and maintained. As single-cell capture and sequencing technology continues to
improve analysis tools will have to adapt to significantly larger datasets (in
the millions of cells) which may require specialised data structures and
algorithms. Other types of single- cell data will become available and projects
such as the Human Cell Atlas [@Regev2017-wg] will provide comprehensive cell
type references, which will open up new avenues for analysis.

As the field expands the scRNA-tools database will continue to be updated. We
hope that it provides a resource for researchers to explore when approaching
scRNA-seq analyses as well as providing a record of the analysis landscape and
how it changes over time.

# Methods

**Database**

When new tools come to our attention they are added to the scRNA-tools database.
DOIs and publication dates are recorded for any associated publications. As
preprints may be frequently updated they are marked as a preprint instead
of recording a date. The platform used to build the tool, links to code
repositories, associated licenses and a short description are also recorded.
Each tool is categorised according to the analysis tasks it can perform,
receiving a true or false for each category based on what is described in the
accompanying paper or documentation. We also record the date that each entry was
added to the database and the date that it was last updated.

**Website**

To build the website we start with the table described above as a CSV file
which is processed using an R script. The lists of packages available in the
CRAN, Bioconductor and PyPI software repositories are downloaded and matched
with tools in the database. For tools with peer-reviewed publications the
number of citations they have received is retrieved from the Crossref database
(www.crossref.org) using the rcrossref package (v0.7.0) [@Chamberlain2017-ov].
JSON files describing the complete table, tools and categories are outputted and
used to populate the website.

The website consists of three main pages. The home page shows an interactive
table with the ability to sort, filter and download the database. The
second page shows an entry for each tool, giving the description, details
of publications, details of the software code and license and the associated
software categories. Badges are added to tools to provide clearly visible
details of any associated software or GitHub repositories. The final page
describes the categories, providing easy access to the tools assoicated with
them.

**Analysis**

The most recent version of the scRNA-tools database was used for the analysis
presented in this paper. Data was manipulated in R using the dplyr package
(v0.7.3) [@Wickham2017-ip] and plots produced using the ggplot2 (v2.2.1)
[@Wickham2010-zq] and cowplot (v0.8.0) [@Wilke2017-zt] packages.

# Declarations

## Ethics

Not applicable.

## Availability of data and materials

The scRNA-tools databases is publicly accessible via the website at www.scRNA-
tools.org. Suggestions for additions, updates and improvements are warmly
welcomed at the associated GitHub repository (https://github.com/Oshlack/scRNA-
tools). The code and datasets used for the analysis in this paper are available
from https://github.com/Oshlack/scRNAtools-paper.

## Competing interests

The authors declare no competing interests.

## Funding

Luke Zappia is supported by an Australian Government Research Training Program
(RTP) Scholarship. Alicia Oshlack is supported through a National Health and
Medical Research Council Career Development Fellowship APP1126157. MCRI is
supported by the Victorian Government's Operational Infrastructure Support
Program.

## Authors' contributions

## Acknowledgements

We would like to acknowledge Sean Davis' work in managing the awesome-single-
cell page and producing a prototype of the script used to process the database.
Daniel Wells had the idea for recording software licenses and provided
licenses for the tools in the database at that time. Breon Schmidt designed a
prototype of the scRNA-tools website and answered many questions about HTML and
Javascript.

# Additional files

# References
